File: Python/20.valid-parentheses.py
#
# @lc app=leetcode id=20 lang=python3
#
# [20] Valid Parentheses
#
class Solution:
    def isValid(self, s: str) -> bool:
        if str == '':
            return True
        if len(s) % 2 != 0:
            return False
        CharDict = {
            '(':1,
            ')':-1,
            '{':10,
            '}':-10,
            '[':100,
            ']':-100
        }

        Result = []
        for i in s:
            if CharDict[i] > 0:
                Result.append(i)
            elif CharDict[i] < 0:
                if len(Result) == 0:
                    return False
                LastChar = Result.pop()
                if CharDict[LastChar] + CharDict[i] != 0:
                    return False
        if len(Result) == 0:
            return True
        return False

File: Java/20.valid-parentheses.java
import java.util.*;

/*
 * @lc app=leetcode id=20 lang=java
 *
 * [20] Valid Parentheses
 */

// @lc code=start
class Solution {
    public boolean isValid(String s) {
        if (s == null) {
            return true;
        }
        if (s.length() % 2 == 1) {
            return false;
        }

        Deque<Character> stack = new LinkedList<>();
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                char top = stack.pop();
                if (top == '(' && c == ')') {
                    continue;
                } else if (top == '[' && c == ']') {
                    continue;
                } else if (top == '{' && c == '}') {
                    continue;
                } else {
                    return false;
                }
            }
        }
        if (!stack.isEmpty()) {
            return false;
        }
        return true;
    }
}
// @lc code=end

